### PR TITLE
gh-79097: Add support for aggregate window functions in sqlite3

### LINKED ISSUE #79097: Add create_window_function() to sqlite3.Connection
BPO | [34916](https://bugs.python.org/issue34916)
--- | :---
Nosy | @ronaldoussoren, @berkerpeksag, @coleifer, @erlend-aasland
PRs | <li>python/cpython#20903</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/berkerpeksag'
closed_at = None
created_at = <Date 2018-10-06.17:29:26.073>
labels = ['extension-modules', 'type-feature', '3.8']
title = 'Add create_window_function() to sqlite3.Connection'
updated_at = <Date 2020-06-15.21:48:40.567>
user = 'https://bugs.python.org/BigStone'
```

bugs.python.org fields:
```python
activity = <Date 2020-06-15.21:48:40.567>
actor = 'erlendaasland'
assignee = 'berker.peksag'
closed = False
closed_date = None
closer = None
components = ['Extension Modules']
creation = <Date 2018-10-06.17:29:26.073>
creator = 'Big Stone'
dependencies = []
files = []
hgrepos = []
issue_num = 34916
keywords = ['patch']
message_count = 8.0
messages = ['327253', '327387', '327429', '327433', '331948', '342206', '349506', '370967']
nosy_count = 7.0
nosy_names = ['ghaering', 'ronaldoussoren', 'berker.peksag', 'Big Stone', 'Scott Stevens', 'coleifer', 'erlendaasland']
pr_nums = ['20903']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue34916'
versions = ['Python 3.8']
```

</p></details>

### PR DESCRIPTION
**Pull request by erlend-aasland**
Fixes #79097

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Rebased onto master and upgraded with `Py_NewRef` and intermingled declarations.

**Comment by erlend-aasland** [comment]
On hold until #23341 is merged

**Comment by erlend-aasland** [comment]
Rebased onto master.

**Comment by erlend-aasland** [comment]
@berkerpeksag Any chance of getting this in before the feature freeze?

**Comment by erlend-aasland** [comment]
Synced with `main` and converted to use the new `callback_context` struct (GH-27456).

**Comment by erlend-aasland** [comment]
Putting this on hold until I've figured out how to get full coverage for the `final` callback.

**Comment by erlend-aasland** [comment]
> Putting this on hold until I've figured out how to get full coverage for the `final` callback.

I finally got around to test this further. It seems like SQLite (as of SQLite 3.38.0) does _not_ propagate errors from the finalize callback of window functions. (Errors in finalize callbacks in normal aggregate functions propagate as expected.) I've adapted the test suite to accommodate for this inconsistency.

**Comment by erlend-aasland** [comment]
TODO:

- [ ] Add tests for `deterministic` keyword
- [ ] Add tests for `directonly` keyword
- [ ] Add tests for `innocuous` keyword

File: Misc/NEWS.d/next/Library/2020-05-24-23-52-03.bpo-40617.lycF9q.rst
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2020-05-24-23-52-03.bpo-40617.lycF9q.rst]:
```suggestion
Add :meth:`sqlite3.Connection.create_window_function` to create aggregate
  window functions in :mod:`sqlite3`. Patch by Erlend E. Aasland.
```

File: Modules/_sqlite/connection.c
Comment (kumaraditya303) [inline Modules/_sqlite/connection.c]:
```suggestion
        static const char *attr_msg = "user-defined aggregate's 'finalize' method "
```

File: Modules/_sqlite/connection.c
Comment (kumaraditya303) [inline Modules/_sqlite/connection.c]:
```suggestion
        static const char *err_msg  = "user-defined aggregate's 'finalize' method "
```

File: Modules/_sqlite/connection.c
Comment (kumaraditya303) [inline Modules/_sqlite/connection.c]:
```suggestion
        static const char *attr_msg = "user-defined aggregate's 'value' method "
```

File: Modules/_sqlite/connection.c
Comment (kumaraditya303) [inline Modules/_sqlite/connection.c]:
```suggestion
        static const char *err_msg  = "user-defined aggregate's 'value' method "
```

File: Misc/NEWS.d/next/Library/2020-05-24-23-52-03.bpo-40617.lycF9q.rst
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2020-05-24-23-52-03.bpo-40617.lycF9q.rst]: I appreciate your comments, Kumar, but I would prefer to get them after I mark my PR ready for review 🙂 Thanks!

**Comment by erlend-aasland** [comment]
@JelleZijlstra, would you like to review this?

FYI, the two buildbot failures look unrelated.

File: Lib/test/test_sqlite3/test_userfunctions.py
Comment (JelleZijlstra) [inline Lib/test/test_sqlite3/test_userfunctions.py:623]: This doesn't actually test that the new definition takes effect, because both window functions do the same thing. Can you make the redefined class do something different than WindowSumInt?

**Review by JelleZijlstra — COMMENTED** [review]
Looks good, just one requested improvement in the tests.

Let me know if there's other changes you'd like to get in in time for 3.11.

File: Lib/test/test_sqlite3/test_userfunctions.py
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_userfunctions.py:623]: Good point. I'll fix that. Thanks!

**Comment by erlend-aasland** [comment]
> Let me know if there's other changes you'd like to get in in time for 3.11.

Yes, I've one more feature I'd like to get in: Blob handles, gh-30680 (currently revisiting it, hoping to make it ready for review soon).

**Comment by JelleZijlstra** [comment]
> Yes, I've one more feature I'd like to get in: Blob handles, [gh-30680](https://github.com/python/cpython/pull/30680) (currently revisiting it, hoping to make it ready for review soon).

Sounds good, ping me when you're ready.

File: Lib/test/test_sqlite3/test_userfunctions.py
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_userfunctions.py:623]: Fixed in 431ace8af2fc93bab6895f6c6128b25f206ab725

**Review by JelleZijlstra — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, @JelleZijlstra! Much appreciated.

**Pull request merged**
